File: python/1079.py
import sys, itertools

sys.stdin = open("1079.txt", 'r')

case = 0
while True:
	n = int(sys.stdin.readline())

	if n == 0:
		break

	case += 1
	a = []
	for i in range(n):
		b = sys.stdin.readline().split()
		a.append((int(b[0])*60, int(b[1])*60))

	b = list(itertools.permutations(a))

	m = -1
	for i in b:
		low, high = 0, 1440 * 60
		p = 0
		while low < high:
			mid = (low + high) // 2
			if mid == low:
				break
			doable = True
			current = 0
			for j in range(len(i)):
				if j == 0:
					current = i[j][0]
				elif j == len(i) - 1:
					if current + mid <= i[j][1]:
						break
					doable = False
				else:
					if i[j][0] > current + mid:
						current = i[j][0]
					elif i[j][1] >= current + mid:
						current += mid
					else:
						doable = False
						break
			if doable:
				low = mid
			else:
				high = mid
		m = max(m, low)
		e = m % 60
		if e < 10:
			e = "0" + str(e)
	print("Case {0}: {1}:{2}".format(case, m // 60, str(e)))

File: c++/1079.cc
#include <bits/stdc++.h>

using namespace std;

int main() 
{
	ifstream cin("1079.txt");

	int c = 0;
	int n;

	while (true) {
		cin >> n;

		if (n == 0) {break;}

		vector<pair<double, double>> a;
		for (int i{0}; i<n; ++i) {
			double s,t;
			cin >> s >> t;
			a.push_back(make_pair(s*60,t*60));
		}
		sort(a.begin(), a.end());

		double m = -1;
		do {
			double low = 0;
			double high = 1440 * 60;
			double best = -1;
			while (fabs(low - high) >= 0.1) {
				double mid = (low + high) / 2.0;
		
				bool doable = true;
				double current = 0;
				for (int j{0}; j<a.size(); ++j) {
					if (j == 0) {
						current = a[j].first;
					} else if (j == a.size() - 1) {
						if (current + mid <= a[j].second) {break;}
						doable = false;
					} else {
						if (a[j].first > current + mid) {
							current = a[j].first;
						} else if (a[j].second >= current + mid) {
							current += mid;
						} else {
							doable = false;
							break;
						}
					}
				}
				if (doable) {
					low = mid;
					best = max(best, mid);
				}
				else {high = mid;}
			}
			m = max(m, best);
		} while (next_permutation(a.begin(), a.end()));
		m = (int)(m + 0.5);
		string s;
		if ((int)m%60 < 10) {
			s = "0" + to_string((int)m%60);
		} else {
			s = to_string((int)m%60);
		}

		cout << "Case " << ++c << ": " << (int)m/60 << ":" << s << endl;

	}
	

}